[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/cindyweng18/csc448_hw/blob/main/Assignment4.ipynb)

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
iris = load_iris()
features = iris['feature_names']
target = 'species' 

df = pd.DataFrame(iris['data'], columns=features)
df[target] = iris['target']

# Output the first 15 rows of the data
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

There are a total of 150 entries for this dataset and the measured numbers for each column (cm) seem to be reasonable and within "normal" range, that is, there isn't a maximum of 20 cm or a minimum of 0 cm. When we print out the `iris` dataset imported from `Sklearn`, we can see that it has the features and labels that we need. The features are `sepal length (cm)`, `sepal width (cm)`, `petal length (cm)`, `petal width (cm)`. And the label is the species that the flower corresponds to which is an integer instead of the string value of the flower species since the target array from the iris dataset is composed of 0, 1, 2. This mapping of labels corresponds to the actual classes: `setosa`, `versicolor`, `virginica` respectively.

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
X = df[features]
Y = df[target]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, Y,test_size=0.1)

# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
reg = LogisticRegression()
reg.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = np.array([[4.4, 4.4, 4.4, 4.4]])
print("Predicted Probabilities [Class 0, Class 1, Class 2]: ", reg.predict_proba(sample)[0])

# iii. Report on the score for Logistic regression model, what does the score measure?
reg_score = reg.score(X_test, y_test)
print("Score: ", reg_score)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coef = reg.coef_
y_int = reg.intercept_
print("Coefficients: ", coef)
print("Y-Intercept: ", y_int)

Predicted Probabilities [Class 0, Class 1, Class 2]:  [1.29486547e-04 7.88544378e-04 9.99081969e-01]
Score:  1.0
Coefficients:  [[-0.40407204  0.92665844 -2.45635453 -1.06431032]
 [ 0.53710156 -0.33020752 -0.20143971 -0.88843883]
 [-0.13302952 -0.59645093  2.65779424  1.95274915]]
Y-Intercept:  [  9.60566957   2.10586786 -11.71153744]


/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


The score for the logistic regression moddel is 1.0, this score measure the mean accuracy of the testing dataset where accuracy is defined as the number of correct predictions made divided by the total number of datapoints in the subset. A score of 1.0 means that the model was able to correctly predict all the datapoints in the testing set with respect to the predicted values obtained from the features testing set and comparing it with the target testing set.

# Part 4: Support Vector Machine

In [6]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = SVC(probability=True)
svc.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = np.array([[4.4, 4.4, 4.4, 4.4]])
print("Predicted Probabilities [Class 0, Class 1, Class 2]:", svc.predict_proba(sample)[0])

# iii. Report on the score for the SVM, what does the score measure?
svc_score = svc.score(X_test, y_test)
print("Score: ", svc_score)

Predicted Probabilities [Class 0, Class 1, Class 2]: [0.07346016 0.03168489 0.89485494]
Score:  1.0


The score for the SVM is 1.0, this score measure the mean accuracy of the testing dataset where accuracy is defined as the number of correct predictions made divided by the total number of datapoints in the subset. A score of 1.0 means that the model was able to correctly predict all the datapoints in the testing set with respect to the predicted values obtained from the features testing set and comparing it with the target testing set.

# Part 5: Neural Network

In [7]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier()
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = np.array([[4.4, 4.4, 4.4, 4.4]])
print("Predicted Probabilities [Class 0, Class 1, Class 2]:", mlp.predict_proba(sample)[0])

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp.score(X_test, y_test)
print("Score: ", mlp_score)

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
diff_mlp = MLPClassifier(
            activation='logistic',
            solver='lbfgs',
            alpha=0.1,
            tol=0.001
)
diff_mlp.fit(X_train, y_train)
diff_score = diff_mlp.score(X_test, y_test)
print("Score of modified neural network: ", diff_score)

Predicted Probabilities [Class 0, Class 1, Class 2]: [7.00672418e-04 3.14370456e-03 9.96155623e-01]
Score:  1.0
Score of modified neural network:  1.0


/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


The score for the neural network is 1.0, this score measure the mean accuracy of the testing dataset where accuracy is defined as the number of correct predictions made divided by the total number of datapoints in the subset. A score of 1.0 means that the model was able to correctly predict all the datapoints in the testing set with respect to the predicted values obtained from the features testing set and comparing it with the target testing set.

# Part 6: K-Nearest Neighbors

In [8]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print("Predicted Probabilities [Class 0, Class 1, Class 2]:", knn.predict_proba(sample)[0])

# iii. Report on the score for kNN, what does the score measure?
knn_score = knn.score(X_test, y_test)
print("Score: ", knn_score)

Predicted Probabilities [Class 0, Class 1, Class 2]: [0.  0.2 0.8]
Score:  1.0


The score for the kNN is 1.0, this score measure the mean accuracy of the testing dataset where accuracy is defined as the number of correct predictions made divided by the total number of datapoints in the subset. A score of 1.0 means that the model was able to correctly predict all the datapoints in the testing set with respect to the predicted values obtained from the features testing set and comparing it with the target testing set.

# Part 7: Conclusions and takeaways
In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results of the notebook were the same througout all the models. The models scored a 100% accuracy so all the models performed best on the dataset. I think this is because this is a small dataset (total of 150 items) and there are very few features that we can use when training a model compared to more realistic datasets where you can have 100,000 items and so many features to choose from when training a model. One thing that surprised me about the exercise was how great the models scored which almost seem like it was too good to be true.